### ISSUE TITLE
Indicate in-place parameter updates in TorchScript IR

### ISSUE THREAD

**Issue by benschreiber** [pytorch/pytorch#41361]
## 🚀 Feature

TorchScript's graph representation does not explicitly indicate when method parameters are modified in-place. It should indicate in the IR where such operations could occur.

## Motivation

TorchScript exists to provide an easy-to-parse, unambiguous representation of a PyTorch model. Its IR graph representation is in SSA form to support this goal. Consider the following program:

```
import torch
import torch.nn.functional as F

def test():
    x = torch.rand((4,3,32,32))
    m = torch.full((3,), 0.0)
    v = torch.full((3,), 1.0)
    result = F.instance_norm(x, running_mean=m, running_var=v, use_input_stats=True)
    return result, m, v

scripted = torch.jit.script(test)
print(scripted.graph)
print(scripted()[1:]) # e.g. (tensor([0.0504, 0.0496, 0.0498]), tensor([0.9083, 0.9082, 0.9083]))
```

The resulting IR graph looks like this:

```
graph():
  %34 : Function = prim::Constant[name="instance_norm"]()
  %33 : float = prim::Constant[value=1.0000000000000001e-05]()
  %32 : float = prim::Constant[value=0.10000000000000001]()
  %29 : bool = prim::Constant[value=1]() # foo.py:8:79
  %5 : None = prim::Constant()
  %0 : int = prim::Constant[value=4]() # foo.py:5:20
  %1 : int = prim::Constant[value=3]() # foo.py:5:22
  %2 : int = prim::Constant[value=32]() # foo.py:5:24
  %11 : float = prim::Constant[value=0.]() # foo.py:6:25
  %19 : float = prim::Constant[value=1.]() # foo.py:7:25
  %4 : int[] = prim::ListConstruct(%0, %1, %2, %2)
  %x.1 : Tensor = aten::rand(%4, %5, %5, %5, %5) # foo.py:5:8
  %12 : int[] = prim::ListConstruct(%1)
  %m.1 : Tensor = aten::full(%12, %11, %5, %5, %5, %5) # foo.py:6:8
  %20 : int[] = prim::ListConstruct(%1)
  %v.1 : Tensor = aten::full(%20, %19, %5, %5, %5, %5) # foo.py:7:8
  %result.1 : Tensor = prim::CallFunction(%34, %x.1, %m.1, %v.1, %5, %5, %29, %32, %33) # foo.py:8:13
  %39 : (Tensor, Tensor, Tensor) = prim::TupleConstruct(%result.1, %m.1, %v.1)
  return (%39)
```

`prim::CallFunction` refers to `F.instance_norm` here. The instance norm function in-place modifies the running mean and running var input tensors in the direction of the computed mean and var of the input. The labels `%m.1` and `%v.1` are returned, even though their values have changed. An external consumer of TorchScript IR may incorrectly optimize around the assumption that `%m.1` and `%v.1` are unchanged.

## Pitch

Functions in TorchScript IR should always return new labels for modified parameters.

## Alternatives

TorchScript could add a mechanism where function signatures could express that a parameter is mutable. Parameters not explicitly marked mutable must be const. 

cc @suo @gmagogsfm

**Comment by eellison** [comment]
Hi, thanks for raising the issue. A few things:

* We have an alias db for analyzing mutation queries. 

If you run 
```
import torch
import torch.nn.functional as F

def test():
    x = torch.rand((4,3,32,32))
    m = torch.full((3,), 0.0)
    v = torch.full((3,), 1.0)
    m.add_(1)
    return x + m + v

scripted = torch.jit.script(test)
scripted.graph.dump_alias_db()

print(scripted.graph)
```
You'll see:
```
===1. GRAPH===
graph():
  %7 : None = prim::Constant()
  %0 : int = prim::Constant[value=4]() # test/elias.py:5:20
  %1 : int = prim::Constant[value=3]() # test/elias.py:5:22
  %2 : int = prim::Constant[value=32]() # test/elias.py:5:24
  %13 : float = prim::Constant[value=0.]() # test/elias.py:6:25
  %22 : float = prim::Constant[value=1.]() # test/elias.py:7:25
  %31 : int = prim::Constant[value=1]() # test/elias.py:8:11
  %6 : int[] = prim::ListConstruct(%0, %1, %2, %2)
  %x.1 : Tensor = aten::rand(%6, %7, %7, %7, %7) # test/elias.py:5:8
  %15 : int[] = prim::ListConstruct(%1)
  %m.1 : Tensor = aten::full(%15, %13, %7, %7, %7, %7) # test/elias.py:6:8
  %24 : int[] = prim::ListConstruct(%1)
  %v.1 : Tensor = aten::full(%24, %22, %7, %7, %7, %7) # test/elias.py:7:8
  %33 : Tensor = aten::add_(%m.1, %31, %31) # test/elias.py:8:4
  %37 : Tensor = aten::add(%x.1, %m.1, %31) # test/elias.py:9:11
  %40 : Tensor = aten::add(%37, %v.1, %31) # test/elias.py:9:11
  return (%40)

===2. ALIAS DB===
%33 points to: %m.1, 

===3. Writes===
%33 : Tensor = aten::add_(%m.1, %31, %31) # test/elias.py:8:4
  %m.1, 
```

Alias Analysis queries are the only way to ensure that something is not being written to. I proposed https://github.com/pytorch/pytorch/issues/39539 to be a little more explicit in our IR printing about what is being written to.

* Almost all TorchScript passes are run after very aggressive inlining, so it's probably unlikely that start doing summaries of function parameters anytime soon... It would not be too hard to reuse alias analysis to do this yourself, but if there's not a particular reason to use inlining I would suggest using that. As of now alias analysis assumes all inputs to a function call are written to. 

* It looks like aten::instance_norm is unfortunately not annotated correctly as writing to its inputs 😞 That's a bug, will file and link to it here. 

Does this answer 
> Functions in TorchScript IR should always return new labels for modified parameters 

or was there something else you were envisioning ?

**Comment by benschreiber** [comment]
Thanks for the detailed response. I will need some time to examine the alias information available and see how it integrates with our system.

aten::instance_norm only writes to those inputs if they are present and use_input_stats is True. If they are present and use_input_stats is False, there should be some way to know that they won't be written to. Will the alias db provide this? If not, is it possible?

I had no idea that the alias db existed. Is there a place where the lower level parts of the JIT API are documented (or at least listed)?

**Comment by eellison** [comment]
It is possible, although it would have to handled specially within alias analysis instead of expressed through the function schema.

This is probably the best place to read documentation, not sure how complete it is: https://github.com/pytorch/pytorch/blob/53af9df557aff745edf24193ece784fd008c6f19/torch/csrc/jit/OVERVIEW.md#alias-analysis-in-the-ir

**Comment by wconstab** [comment]
@benschreiber We've filed #42317 to track the instance norm annotation.  Can you respond if there are other specific issues here, or otherwise close this out?  Thanks!
